NETOBSERV-2071 CLI 1.8 improvments (#5)
netobserv/netobserv.github.io: content/posts/2025-01-23-cli_whats_new_1.8.md

---
layout: :theme/post
title: "Network Observability On Demand 1.8 Update"
description: Command line interface improvements in version 1.8
tags: CLI,Monitoring,Troubleshooting
authors: [jpinsonneau]
---

> # Network Observability On Demand 1.8 Update
> 
> ```

Author: ```suggestion
# Network Observability On Demand 1.8 Update

_Thanks to: Joël Takvorian for reviewing_

```
```

         _  _     _       _                       ___ _    ___
        | \| |___| |_ ___| |__ ___ ___ _ ___ __  / __| |  |_ _|
        | .' / -_)  _/ _ \ '_ (_-</ -_) '_\ V / | (__| |__ | | 
        |_|\_\___|\__\___/_.__/__/\___|_|  \_/   \___|____|___|

------------------------------------------------------------------------
```

Since we [introduced the Network Observability CLI](./2024-07-25-cli.md), numerous features have been added. This article will cover the improvements in version 1.8 and provide some concrete examples.

## New Options Available
This update adds several options to the CLI, covering more scenarios and enabling scripting on top of the tool.

### Operate in the Background
The `--background` option allows you to start a flow or packet capture without connecting your terminal to the collector pod. This enables you to let the capture run in the background while you work on something else. You can check the capture's progress using the `follow` command and copy the collector output locally using the `copy` command. Once the job is complete, you can `stop` or `cleanup` everything.

```sh
oc netobserv flows --background # Run a flow capture in the background

> oc netobserv follow             # Show current capture progression

Reviewer: ```suggestion
oc netobserv follow             # Show current capture progress
```

> oc netobserv stop               # Stop the capture keeping the collector running

Reviewer: Not sure if it will be totally clear to the users when reading "... keeping the collector running", they might think collection can continue, which contradicts the beginning of the sentence. Maybe replace with something like:

```
# Stop the capture, keeping the collector pod alive
```
?

Author: sure sounds good to me

```suggestion
oc netobserv stop               # Stop the capture, keeping the collector pod alive
```

oc netobserv copy               # Copy the output

oc netobserv cleanup            # Remove all
```

### Customizable Namespace
You can now customize the capture namespace using the `NETOBSERV_NAMESPACE` environment variable. When the CLI starts, it automatically checks if this namespace exists and will stop if it finds any conflict with a pending capture. This is particularly useful if you want to run captures in parallel.

```sh
NETOBSERV_NAMESPACE=my_ns oc netobserv [flows|packets|metrics|follow|stop|copy|cleanup]
```

> ### Subnets Labelling

Author: ```suggestion
### Subnets Labelling (for OCP clusters)
```

> The tool can now read configurations from `cluster-config-v1` and `network` to identify **Machine**, **Pods**, and **Services** subnets using the `--get-subnets` option. This will automatically add `SrcSubnetLabel` and `DstSubnetLabel` to your flows.

Reviewer: Maybe mention `in OpenShift` ? the `network` CR only exists in openshift

Author: Sure, I can put that in the title and the text:

```suggestion
The tool can now read Openshift configurations from `cluster-config-v1` and `network` to identify **Machine**, **Pods**, and **Services** subnets using the `--get-subnets` option. This will automatically add `SrcSubnetLabel` and `DstSubnetLabel` to your flows.
```

You will see subnets being configured during the creation of the agents:
```sh
creating flow-capture agents:
opt: get_subnets, value: true
Found subnets:
    Services: "172.30.0.0/16"
    Pods: "10.128.0.0/14"
    Machines: "10.0.0.0/16"
```

![subnet labels]({page.image('cli-whats-new-1-8/subnets.png')})

## Enhanced Data Filtering
Filtering is crucial to gather precise network data without involving excessive resources and storage. The CLI focuses on this area, allowing you to deploy agents only where needed and fine-tune what's captured.

### Assigning Agents to Nodes
It's now possible to define agents `nodeSelector` to capture on a subset of nodes. You can rely on existing labels or create a dedicated one for this usage. For example, you can run:
```sh
oc netobserv flows --node-selector=kubernetes.io/hostname:my-node
```
to run the agents on the node with the `kubernetes.io/hostname:my-node` label.

### eBPF Filters

> Agents recently introduced [the ability to filter](https://github.com/netobserv/netobserv-ebpf-agent/blob/main/docs/flow_filtering.md) on IPs, Ports, Protocol, Action, TCPFlags, and more simultaneously. You can now apply these filters in the CLI as shown below:

Reviewer: Maybe, to be done later, link to @msherif1234 's blog on filters when it ready

Author: ```suggestion
Agents recently introduced [the ability to filter](./2025-01-30-ebpf-flows-filtering.md) on IPs, Ports, Protocol, Action, TCPFlags, and more simultaneously. You can now apply these filters in the CLI as shown below:
```

```sh
oc netobserv flows \              # Capture flows
--protocol=TCP --port=8080 \      # either on TCP 8080
or --protocol=UDP                 # or UDP
```

You will see filters being validated during the creation of the agents:
```sh
creating flow-capture agents:
opt: filter_protocol, value: TCP
opt: filter_port, value: 8080
opt: add_filter
opt: filter_protocol, value: UDP
```

![filters]({page.image('cli-whats-new-1-8/filters.png')})

> You can add as many filters as you want and separate them by or to create multiple capture scenarios.

Reviewer: I guess it's not as many as you want, in the agent it is limited to 16, so I think the same limit exists here.
BTW - @msherif1234 - this limit of 16 seems to be documented on the [operator](https://github.com/netobserv/network-observability-operator/blob/main/apis/flowcollector/v1beta2/flowcollector_types.go#L303) side but not in the agent doc ( https://github.com/netobserv/netobserv-ebpf-agent/blob/main/docs/flow_filtering.md ), I guess we should add that

Author: ```suggestion
You can add up to 16 sets of filters separated by `or`operator to create multiple capture scenarios.
```

Author: Would that work for you @jotak ? :point_up:

Reviewer: yes except for the missing space after `or` :-)

### Regular Expressions Usage

> If you need to filter on enriched content beyond the agent-level filters, you can use **regexes** to match any field/value pair. To filter all traffic from OpenShift namespaces, for example, you can use `--regexes=SrcK8S_Namespace~openshift.*`.

Author: ```suggestion
If you need filtering in on enriched content beyond the agent-level filters, you can use **regexes** to match any field/value pair. To keep only the traffic from OpenShift namespaces, for example, you can use `--regexes=SrcK8S_Namespace~openshift.*`.
```

like this ? :point_up:

> If you need to filter on enriched content beyond the agent-level filters, you can use **regexes** to match any field/value pair. To filter all traffic from OpenShift namespaces, for example, you can use `--regexes=SrcK8S_Namespace~openshift.*`.

Reviewer: yes, or maybe if you keep just the second change "To keep only the traffic..." it's sufficient as it clarifies the operation

> If you need to filter on enriched content beyond the agent-level filters, you can use **regexes** to match any field/value pair. To filter all traffic from OpenShift namespaces, for example, you can use `--regexes=SrcK8S_Namespace~openshift.*`.

Author: sure

You will see regexes being validated during the creation of the agents:
```sh
creating flow-capture agents:
opt: filter_regexes, value: SrcK8S_Namespace~openshift.*
key: SrcK8S_Namespace value: openshift.*
```

![regexes]({page.image('cli-whats-new-1-8/regexes.png')})

Regexes are comma-separated, so you can use multiple at once, such as `--regexes=SrcK8S_Namespace~my-ns,SrcK8S_Name~my-app`. Refer to the [flows format](https://github.com/netobserv/network-observability-operator/blob/main/docs/flows-format.adoc) to see the possible fields.

## Unified Collector UI

> Capturing **packets** now resembles **flows** capture, allowing you to live filter the content. This improvement was made possible by introducing the [flowlogs-pipeline](https://github.com/netobserv/flowlogs-pipeline) component inside [eBPF agents](https://github.com/netobserv/netobserv-ebpf-agent), which parse packets and generate flows from them. All filtering capabilities are compatible with this approach!

Reviewer: ```suggestion
Capturing **packets** now resembles **flows** capture, allowing you to live filter the content. This improvement was made possible by introducing the [flowlogs-pipeline](https://github.com/netobserv/flowlogs-pipeline) component inside [eBPF agents](https://github.com/netobserv/netobserv-ebpf-agent), which parses packets and generates flows from them. All filtering capabilities are compatible with this approach!
```

> ## Unified Collector UI
> Capturing **packets** now resembles **flows** capture, allowing you to live filter the content. This improvement was made possible by introducing the [flowlogs-pipeline](https://github.com/netobserv/flowlogs-pipeline) component inside [eBPF agents](https://github.com/netobserv/netobserv-ebpf-agent), which parses packets and generates flows from them. All filtering capabilities are compatible with this approach!
> 

Reviewer 2: ```suggestion
## Unified Collector UI
Capturing **packets** now resembles **flows** capture, allowing you to live filter the content. This improvement was made possible by introducing the [flowlogs-pipeline](https://github.com/netobserv/flowlogs-pipeline) component inside [eBPF agents](https://github.com/netobserv/netobserv-ebpf-agent), which parses packets and generates flows from them. All filtering capabilities are compatible with this approach!
```

I would start with the last sentence first `All filtering capabilities are now supported for packet capture option as well.` since this is huge benefit and and highlight heading to be something like `Filtering with Packet captures` if filtering capabilities is only thing we want to highlight here.

Author: Would that work better ? 
https://github.com/netobserv/netobserv.github.io/pull/5/commits/26eaa907908719b4976475e1618e0e24725ed840

Reviewer 2: looks good.

Author: thanks !

> Run a packet capture on a specitic port for example:

Reviewer: ```suggestion
Run a packet capture on a specific port for example:
```

```sh
oc netobserv packets --port=80
```

![packet capture]({page.image('cli-whats-new-1-8/packets.png')})

## Metrics Capture on OpenShift
Capturing metrics is now possible using the `metrics` command. This creates a `ServiceMonitor` to gather metrics from the agents and store them in [Prometheus](https://prometheus.io/). You can enable all or specific features to gather more information about your network, such in:
```sh
oc netobserv metrics --enable_all
``` 
to capture packet drops, DNS and RTT metrics or 
```sh
oc netobserv metrics --enable_pktdrop
```
to focus only on drops.

> On top of the features, you can use all the filtering capabilities mentioned above to focus on what you're looking for. A `NetObserv / On Demand` dashboard will be automatically created, showing the results.

Author: The metrics capture never stops by iteself. You need to run the stop or cleanup command.

I can add a sentence about that after the screenshot

> On top of the features, you can use all the filtering capabilities mentioned above to focus on what you're looking for. A `NetObserv / On Demand` dashboard will be automatically created, showing the results.

Author: https://github.com/netobserv/netobserv.github.io/pull/5/commits/1de7c5ae0cdfb1ade072692379c98f04f8b70e39

oc netobserv copy               # Copy the output

oc netobserv cleanup            # Remove all
```

### Customizable Namespace
You can now customize the capture namespace using the `NETOBSERV_NAMESPACE` environment variable. When the CLI starts, it automatically checks if this namespace exists and will stop if it finds any conflict with a pending capture. This is particularly useful if you want to run captures in parallel.

```sh
NETOBSERV_NAMESPACE=my_ns oc netobserv [flows|packets|metrics|follow|stop|copy|cleanup]
```

### Subnets Labelling
The tool can now read configurations from `cluster-config-v1` and `network` to identify **Machine**, **Pods**, and **Services** subnets using the `--get-subnets` option. This will automatically add `SrcSubnetLabel` and `DstSubnetLabel` to your flows.

You will see subnets being configured during the creation of the agents:
```sh
creating flow-capture agents:
opt: get_subnets, value: true
Found subnets:
    Services: "172.30.0.0/16"
    Pods: "10.128.0.0/14"
    Machines: "10.0.0.0/16"
```

![subnet labels]({page.image('cli-whats-new-1-8/subnets.png')})

### YAML Configuration Output

> Outputting a `.yml` file instead of running the capture is now possible using the `--yaml` option. The file will contain all the resources needed to run the capture, such as the namespace, the agents embedding the pipeline and its configuration, and the related services. The collector will need to be run manually in parallel to start the capture.

Author: That may not make it for 1.8 depending on https://github.com/netobserv/network-observability-cli/pull/148

Author: removed for now: https://github.com/netobserv/netobserv.github.io/pull/5/commits/b1d103ecf52e8ac516a66c5b4c049364e047d6a6

---
layout: :theme/post
title: "Network Observability On Demand 1.8 Update"
description: Command line interface improvements in version 1.8
tags: CLI,Monitoring,Troubleshooting
authors: [jpinsonneau]
---

# Network Observability On Demand 1.8 Update

_Thanks to: Joël Takvorian, Mehul Modi and Sara Thomas for reviewing_

> 

Author: ```suggestion

```
```

## Metrics Capture on OpenShift
Capturing metrics is now possible using the `metrics` command. This creates a `ServiceMonitor` to gather metrics from the agents and store them in [Prometheus](https://prometheus.io/). You can enable all or specific features to gather more information about your network, such in:
```sh
oc netobserv metrics --enable_all
``` 
to capture packet drops, DNS and RTT metrics or 
```sh
oc netobserv metrics --enable_pktdrop
```
to focus only on drops.

On top of the features, you can use all the filtering capabilities mentioned above to focus on what you're looking for. A `NetObserv / On Demand` dashboard will be automatically created, showing the results.

![dashboard]({page.image('cli-whats-new-1-8/dashboard.png')})

In this mode, the maximum bytes or time options are ignored since there is no collector pod involved. Only the eBPF agents are deployed. 
You will need to run `stop` or `cleanup` command to remove everything.

## Help!

The help has been enhanced to provide examples for each command and option. You can type `oc netobserv help` for the general help message:

```sh
$ oc netobserv help

> Netobserv allows you to capture flows, packets and metrics from your cluster.

Reviewer 3: ```suggestion
Netobserv CLI allows you to capture flows, packets and metrics from your cluster.
```

    --protocol=TCP --port=8080 \                              # either on TCP 8080
    or --protocol=UDP                                         # or UDP
    netobserv follow                                          # Display the progression of the background capture
    netobserv stop                                            # Stop the background capture by deleting eBPF agents
    netobserv copy                                            # Copy the background capture output data
    netobserv cleanup                                         # Cleanup netobserv CLI by removing the remaining collector pod

  Capture packets on specific nodes and port
    netobserv packets                                         # Capture packets
    --node-selector=netobserv:true \                          # on nodes labelled with netobserv=true
    --port=80 \                                               # on port 80 only
    --max-bytes=100000000                                     # for a maximum of 100MB
```

You can also request help on a specific command, such as `oc netobserv metrics help`, to get its options list:
```sh
$ oc netobserv metrics help

> Netobserv allows you to capture metrics on your OCP cluster.

Reviewer 3: ```suggestion
Netobserv CLI allows you to capture metrics on your OCP cluster.
```

  --peer_ip:                    filter peer IP                             (default: n/a)
  --peer_cidr:                  filter peer CIDR                           (default: n/a)
  --port_range:                 filter port range                          (default: n/a)
  --port:                       filter port                                (default: n/a)
  --ports:                      filter on either of two ports              (default: n/a)
  --protocol:                   filter protocol                            (default: n/a)
  --regexes:                    filter flows using regular expression      (default: n/a)
  --sport_range:                filter source port range                   (default: n/a)
  --sport:                      filter source port                         (default: n/a)
  --sports:                     filter on either of two source ports       (default: n/a)
  --tcp_flags:                  filter TCP flags                           (default: n/a)
  --interfaces:                 interfaces to monitor                      (default: n/a)
```

## Feedback
We hope you enjoyed this article!

> Netobserv is an OpenSource project [available on github](https://github.com/netobserv).

Reviewer: nit: open source is 2 words

```suggestion
Netobserv is an open source project [available on github](https://github.com/netobserv).
```
